lib/spack/spack/cmd/ci.py

```diff
@@ -304,82 +347,63 @@ def ci_rebuild(args):
-                # file on the remote mirror.
-                first_pass_args = install_args + [
-                    '--cache-only',
-                    '--only',
-                    'dependencies',
-                ]
-                first_pass_args.extend(spec_cli_arg)
-                tty.debug('First pass install arguments: {0}'.format(
-                    first_pass_args))
-                spack_cmd(*first_pass_args)
-
-                # Overwrite the changed environment file so it doesn't break
-                # the next install invocation.
-                tty.debug('Copying {0} to {1}'.format(
-                    env_dst_path, env_src_path))
-                shutil.copyfile(env_dst_path, env_src_path)
-
-                second_pass_args = install_args + [
-                    '--no-cache',
-                    '--only',
-                    'package',
-                ]
-                second_pass_args.extend(cdash_args)
-                second_pass_args.extend(spec_cli_arg)
-                tty.debug('Second pass install arguments: {0}'.format(
-                    second_pass_args))
-                spack_cmd(*second_pass_args)
-            except Exception as inst:
-                tty.error('Caught exception during install:')
-                tty.error(inst)
+                tty.debug('spack install arguments: {0}'.format(
+                    install_args))
+                spack_cmd(*install_args)
+            finally:
+                spack_ci.copy_stage_logs_to_artifacts(job_spec, job_log_dir)
+
+            # Create buildcache on remote mirror, either on pr-specific
+            # mirror or on mirror defined in spack environment
+            if spack_is_pr_pipeline:
+                buildcache_mirror_url = pr_mirror_url
+            else:
+                buildcache_mirror_url = remote_mirror_url
 
-            spack_ci.copy_stage_logs_to_artifacts(job_spec, job_log_dir)
-
-            # 4) create buildcache on remote mirror, but not if this is
-            # running to test a spack PR
-            if not spack_is_pr_pipeline:
+            try:
                 spack_ci.push_mirror_contents(
-                    env, job_spec, job_spec_yaml_path, remote_mirror_url,
-                    cdash_build_id)
-
-            # 5) create another copy of that buildcache on "local artifact
-            # mirror" (only done if cash reporting is enabled)
+                    env, job_spec, job_spec_yaml_path, buildcache_mirror_url,
+                    cdash_build_id, sign_binaries)
+            except Exception as inst:
+                # If the mirror we're pushing to is on S3 and there's some
+                # permissions problem, for example, we can't just target
+                # that exception type here, since users of the
+                # `spack ci rebuild' may not need or want any dependency
+                # on boto3.  So we use the first non-boto exception type
+                # in the heirarchy:
```

Comment:
I wonder if it makes sense to shim out the few Exception classes needed from boto?  I'll leave it up to you to decide if the maintenance cost is worth it.

Reply:
I'm not sure it's worth the maintenance cost, but I could be convinced, maybe I'm just not seeing the value.